server/app/models/grid_service.rb

```diff
@@ -220,6 +220,20 @@ def overlay_network?
     self.net.to_s == 'bridge'
   end
 
+  # @param link [String] service, stack/service, /service
+  # @return [GridService, nil]
+  def resolve_service(link)
```

Comment:
Maybe this method should be within affinity filter.. it's the only place where it's used?

Reply:
There's something very similar going on in [`GridCertificates::Common#resolve_service`](https://github.com/kontena/kontena/blob/eb00568ccd013406be354dd24f30b5c2a3453a9f/server/app/mutations/grid_certificates/common.rb#L50-L56) used for domain authorization linked services, the [grid `.../container_logs` API](https://github.com/kontena/kontena/blob/eb00568ccd013406be354dd24f30b5c2a3453a9f/server/app/routes/v1/grids_api.rb#L140-L148), the `GridServices::Common` mutation ([build](https://github.com/kontena/kontena/blob/eb00568ccd013406be354dd24f30b5c2a3453a9f/server/app/mutations/grid_services/common.rb#L36-L41) and [validate](https://github.com/kontena/kontena/blob/eb00568ccd013406be354dd24f30b5c2a3453a9f/server/app/mutations/grid_services/common.rb#L160-L168))... it would be nice to have a single place to handle those kind of `stack/service` references, and IMO that would belong in the models. Unfortunately each of those uses is also slightly different, so it's not appropriate to try and refactor all of them in this PR.

I guess I'll move this into the affinity filter for now, and leave it up to a later refactoring PR to try and consolidate all of those implementations of the same thing.